server/src/main/kotlin/com/xebia/functional/xef/server/http/routes/Routes.kt

```diff
@@ -5,65 +5,124 @@ import com.aallam.openai.api.chat.ChatCompletionRequest
+import kotlinx.serialization.json.jsonPrimitive
 
 enum class Provider {
-  OPENAI, GPT4ALL, GCP
+    OPENAI, GPT4ALL, GCP
 }
 
 fun String.toProvider(): Provider? = when (this) {
-  "openai" -> Provider.OPENAI
-  "gpt4all" -> Provider.GPT4ALL
-  "gcp" -> Provider.GCP
-  else -> null
+    "openai" -> Provider.OPENAI
+    "gpt4all" -> Provider.GPT4ALL
+    "gcp" -> Provider.GCP
+    else -> null
 }
 
 
 @OptIn(BetaOpenAI::class)
-fun Routing.routes(persistenceService: PersistenceService) {
-  authenticate("auth-bearer") {
-    post("/chat/completions") {
-      val provider: Provider = call.request.headers["xef-provider"]?.toProvider()
-        ?: throw IllegalArgumentException("Not a valid provider")
-      val token = call.principal<UserIdPrincipal>()?.name ?: throw IllegalArgumentException("No token found")
-      val scope = Conversation(
-        persistenceService.getVectorStore(provider, token)
-      )
-      val data = call.receive<ChatCompletionRequest>().toCore()
-      val model: OpenAIModel = data.model.toOpenAIModel(token)
-      response<String, Throwable> {
-        model.promptMessage(
-          prompt = Prompt(
-            messages = data.messages,
-            configuration = PromptConfiguration(
-              temperature = data.temperature,
-              numberOfPredictions = data.n,
-              user = data.user ?: ""
-            )
-          ),
-          scope = scope
-        )
-      }
+fun Routing.routes(
+    client: HttpClient,
+    persistenceService: PersistenceService
+) {
+    val openAiUrl = "https://api.openai.com/v1"
+
+    authenticate("auth-bearer") {
+        post("/chat/completions") {
+            val provider: Provider = call.getProvider()
+            val token = call.getToken()
+            val scope = Conversation(persistenceService.getVectorStore(provider, token))
+            val darta = call.receive<String>()
+            val data = Json.decodeFromString<JsonObject>(darta)// call.receive<JsonObject>()
+            if (!data.containsKey("model")) {
+                call.respondText("No model found", status = HttpStatusCode.BadRequest)
+                return@post
+            }
+            val model: OpenAIModel = data["model"]?.jsonPrimitive?.content?.toOpenAIModel(token) ?: run {
+                call.respondText("No model found", status = HttpStatusCode.BadRequest)
+                return@post
+            }
+
+            println(darta)
+
+            val isStream = data["stream"]?.jsonPrimitive?.boolean ?: false
+
+            if (!isStream) {
+                val response = client.request("$openAiUrl/chat/completions") {
+                    headers {
+                        bearerAuth(token)
+                    }
+                    contentType(ContentType.Application.Json)
+                    method = HttpMethod.Post
+                    setBody(darta)
+                }
+                call.respond(response.body<String>())
+            } else {
+                runBlocking {
```

Comment:
why do we need to call `runBlocking` here? 

I think we should be able to get the response straight from OpenAI and use `call.respondOutputStream(openAIResponse.outputStream)` or a similar API where we just connect the two output streams.

We are manually unpacking the bytes and buffering them with a different flush and buffer policy than the one OpenAI is using, so we may not stream back the chunks like they do if we do it like this.